python/tvm/relay/op/contrib/dnnl.py

```diff
@@ -207,7 +211,207 @@ def pattern_table():
+    elif conv_type == 2:
+        data_dic = {"a": "N", "b": "C", "c": "H", "d": "W"}
+        weight_dic = {"a": "O", "b": "I", "c": "H", "d": "W"}
+        if "e" in input_data:
+            weight_dic = {"a": "G", "b": "O", "c": "I", "d": "H", "e": "W"}
+    elif conv_type == 3:
+        data_dic = {"a": "N", "b": "C", "c": "D", "d": "H", "e": "W"}
+        weight_dic = {"a": "O", "b": "I", "c": "D", "d": "H", "e": "W", "f": "G"}
+
+    dic = weight_dic if is_weight else data_dic
+    res = ""
+
+    for i in input_data:
+        if i.isupper():
+            i = i.lower()
+            res += dic[i]
+            dic[i] = dic[i].lower()
+        elif i.islower():
+            res += dic[i]
+        elif i.isdigit():
+            res += i
+        else:
+            raise ValueError("Unsupport layout format: %s" % input_data)
+    return res
+
+
+def legalize_group_conv(attrs, inputs, types):
+    """Legalize group conv's calculation.
+    Alter weight layout from OIHW to GOIHW"""
+    G = attrs.groups
+    if G == 1:
+        return
+    data, weight = inputs
+    OC, IC, H, W = get_shape(weight)
+    new_attrs = dict(attrs)
+    weight = relay.reshape(weight, (G, OC // G, IC, H, W))
+    new_attrs["kernel_layout"] = "GOIHW"
+    return relay.nn.conv2d(data, weight, **new_attrs)
+
+
+def legalize_group_deconv(attrs, inputs, types):
+    """Legalize group deconv's calculation.
+    Alter weight layout from IOHW to GIOHW"""
+    G = attrs.groups
+    if G == 1:
+        return
+    data, weight = inputs
+    IC, OC, H, W = get_shape(weight)
+    new_attrs = dict(attrs)
+    new_attrs["kernel_layout"] = "GIOHW"
+    weight = relay.reshape(weight, (G, IC // G, OC, H, W))
+    return relay.nn.conv2d_transpose(data, weight, **new_attrs)
+
+
+def alter_conv(attrs, inputs, tinfos, out_type):
+    """The convolution's layout auto-query func for dnnl."""
+
+    data, weight = inputs
+    G = str(attrs.groups)
+    weight_shape = ",".join([str(x) for x in get_shape(weight)])
+    out_shape = ",".join([str(x) for x in get_shape(out_type)])
+    paddings = ",".join([str(x) for x in attrs.get_int_tuple("padding")])
+    strides = ",".join([str(x) for x in attrs.get_int_tuple("strides")])
+    dilates = ",".join([str(x) for x in attrs.get_int_tuple("dilation")])
+    new_attrs = dict(attrs)
+    conv_type = len(get_shape(out_type)) - 2
+
+    res = get_optimal_layout_for_conv(
+        len(get_shape(out_type)), weight_shape, out_shape, paddings, strides, dilates, G
+    )
+    src_df, weight_df, dst_df = res.split(",")
+    new_attrs["data_layout"] = validate_layout_for_tvm(src_df, is_weight=False, conv_type=conv_type)
+    new_attrs["kernel_layout"] = validate_layout_for_tvm(
+        weight_df, is_weight=True, conv_type=conv_type
+    )
+    new_attrs["out_layout"] = validate_layout_for_tvm(dst_df, is_weight=False, conv_type=conv_type)
+    if new_attrs["kernel_layout"] == "HWOIG16g":
+        new_attrs["kernel_layout"] = "HWIOG16g"
```

Comment:
Better deal with similar cases with a more general treatment.

Reply:
This is a corner case, I will try to figure out the reason for querying incorrect layout.

Reply:
This layout can now be removed from this PR.

Reply:
nice